### DISCUSSION TITLE
License format error

### DISCUSSION THREAD

**Discussion by dmbrtlt** [totumonline/totum-ru-issues-and-discussions#625] [Как это сделать?] [answered]
Подскажите пожалуйста с лицензиями PRO. На локальный хост (192.168.100.150)  такая ошибка License format error ... No host in license to check

**Comment by ttm-alexeypolunin** [comment]
Ок, посмотрим сегодня.

**Reply by dmbrtlt** [reply]
Перестало вообще пускать, на сайт...
GOMODULE: Schema totum not in license

**Comment by ttm-alexeypolunin** [comment, marked as answer]
Мы проверили. ip в качестве хоста проходит ок

Тестовая лицензия (`"type":"test"`):
```
[{"schemas":{"80.87.107.237":"totum"},"to":"2025-11-15","type":"test","users":20}]
---------
GQrrKPHM0H2fQr03Ka8/KYWC....
```

Записывается в файл `PRO.license-test`:

```
root@vm-c13d738f:~# su totum

totum@vm-c13d738f:/root$ cd /home/totum/totum-mit/

totum@vm-c13d738f:~/totum-mit$ nano PRO.license-test

totum@vm-c13d738f:~/totum-mit$ ls -lh

......
......
-rw-rw-r-- 1 totum totum  606 Nov 15 13:21 PRO.license-test
......
......

totum@vm-c13d738f:~/totum-mit$ bin/gomodule-test

2024-11-15 13:22:57
Opening Conf by pathConf.php: OK
Connecting db totum: OK
Checking License file:
[Read License file PRO.license-test]
[select id, name->>'v' as name from "totum".tables where name->>'v' in ('tables', 'users', 'tables_fields', 'notifications', 'roles')]
[License loaded successfully]
Checking Licenses hosts:
[Part  80.87.107.237 & totum passed successfully]
```

Обратите внимание, что тестовая лицензия записывается в `PRO.license-test`, делается это от пользователя totum.

Для работы лицензий указывается тот хост, по которому осуществляется доступ. Этот же хост должен быть доступен с самого сервера!

Т.е. при выполнение с сервера `curl -v -L HOST` (HOST надо заменить на ваш) вы должны увидеть html предупреждения totum о том, что для работы необходимо включить JS.

**Тестовая лицензия, служит только для проверки корректности хоста и схемы БД и не дает никакого тестового периода. Эта лицензия НЕ позволяет создавать пользователей в системе сверх бесплатного лимита.**

После проверки хоста при полученном `[Part  HOST & SCHEMA passed successfully]` — подтверждаете в ЛК, получаете счет, оплачиваете, деньги зачисляются как приходят на РС и после этого можно активировать рабочую лицензию.

Срок начинается с момента активации.

Рабочая лицензия (`"type":"base"`):

```
[{"schemas":{"80.87.107.237":"totum"},"to":"2025-11-15","type":"base","users":20}]
---------
XHBRR+7gdnLbnk/3V6L7DYIVz
```

Записывается в файл `PRO.license` и после этого необходимо рестартовать GOM:

```
root@vm-c13d738f:~# su totum

totum@vm-c13d738f:/root$ cd /home/totum/totum-mit/

totum@vm-c13d738f:~/totum-mit$ nano PRO.license

totum@vm-c13d738f:~/totum-mit$ service totum-gom restart
Failed to restart totum-gom.service: Access denied
See system logs and 'systemctl status totum-gom.service' for details.

totum@vm-c13d738f:~/totum-mit$ sudo service totum-gom restart
```

Обратите внимание, что от пользователя totum перезапуск выполняется только через sudo.

**Эта лицензия позволяет создавать пользователей в системе сверх бесплатного лимита в том лимите, который оплачен и указан в лицензии!**

Можно удостоверится, что рабочая лицензия прочиталась верно, запустив GOM в режиме отладки.

Перейти в root и остановить gom:

```
totum@vm-c13d738f:~/totum-mit$ exit
exit

root@vm-c13d738f:~# service totum-gom stop
```

Далее от totum запускаем GOM с отладкой лицензий:

```
root@vm-c13d738f:~# su totum

totum@vm-c13d738f:/root$ cd /home/totum/totum-mit/

totum@vm-c13d738f:~/totum-mit$ /home/totum/totum-mit/bin/gomodule -php-sessions-directory /var/lib/php/sessions_totum -totum-directory /home/totum/totum-mit/ -log=ln

13:50:43 LICENSE:[Read License file /home/totum/totum-mit/PRO.license]
13:50:43 LICENSE:[License loaded successfully]
13:50:43 LICENSE:[Check licensePart]
13:52:51 LICENSE:[11  users&connections in schema  totum]
13:52:51 LICENSE:[Clear blocking all schemas in licensePart  map[totum:true]]
13:52:51 LICENSE:[Clear error  totum]
```

Далее можно остановить и запустить gom как сервис:

```
totum@vm-c13d738f:~/totum-mit$ sudo service totum-gom restart
```

Пользователи в логе GOM `[11  users&connections in schema  totum]` указываются по их слотам и без учета технических слотов.

Если записать тестовую лицензию в файл рабочей и попытаться запустить GOM, то он выдаст ошибку:

```
totum@vm-c13d738f:~/totum-mit$ /home/totum/totum-mit/bin/gomodule -php-sessions-directory /var/lib/php/sessions_totum -totum-directory /home/totum/totum-mit/ -log=ln

13:57:54 LICENSE:[Read License file /home/totum/totum-mit/PRO.license]
License type test is not supported
```

PRO-версия до 10 пользователей не проверяет файл лицензии. За пользователя в том числе считаются слоты для одновременных подключений.

Если создать пользователей сверх лимита в действующей лицензии или сверх 10 при отсутствии лицензии (_тестовые лицензии не влияют никак на работу системы, они нужны только для проверки корректности хостов_), то будет выдана ошибка:

```
GOMODULE: Schema totum not in license.
```


Что бы из консоли посмотреть список включенных пользователей:

```
root@vm-c13d738f:~# su totum

totum@vm-c13d738f:/root$ cd /home/totum/totum-mit/

totum@vm-c13d738f:~/totum-mit$ bin/totum schema-users --schema totum --list=on
+---------+---------------+-------+--------+----------+----------+------------+
| Login   | FIO           | Email | Status | Licenses | Tables   | Limit User |
+---------+---------------+-------+--------+----------+----------+------------+
| admin   | Администратор |       | ON     |          | 50       |            |
| cron    | Cron          |       | ON     |          | 0        |            |
| service | service       |       | ON     |          | 50       |            |
| test1   | test1         |       | ON     |          | 10       |            |
| test10  | test10        |       | ON     |          | 10       |            |
| test11  | test11        |       | ON     |          | 10       |            |
| test2   | test2         |       | ON     |          | 10       |            |
| test3   | test3         |       | ON     |          | 10       |            |
| test4   | test4         |       | ON     |          | 10       |            |
| test5   | test5         |       | ON     |          | 10       |            |
| test6   | test6         |       | ON     |          | 0        | +          |
| test7   | test7         |       | ON     |          | 0        | +          |
| test8   | test8         |       | ON     |          | 0        | +          |
| test9   | test9         |       | ON     |          | 0        | +          |
|         |               |       | ALL:14 | FULL:7   | LIMIT:4  | TOTAL:11   |
+---------+---------------+-------+--------+----------+----------+------------+
```

Что бы отключить пользователя:

```
totum@vm-c13d738f:~/totum-mit$ bin/totum schema-users --schema totum --off=test9
```

Что бы укладываться в бесплатный лимит, TOTAL должен быть меньше или равен 10:

```
totum@vm-c13d738f:~/totum-mit$ bin/totum schema-users --schema totum --list=on
+---------+---------------+-------+--------+----------+----------+------------+
| Login   | FIO           | Email | Status | Licenses | Tables   | Limit User |
+---------+---------------+-------+--------+----------+----------+------------+
| admin   | Администратор |       | ON     |          | 50       |            |
| cron    | Cron          |       | ON     |          | 0        |            |
| service | service       |       | ON     |          | 50       |            |
| test1   | test1         |       | ON     |          | 10       |            |
| test10  | test10        |       | ON     |          | 10       |            |
| test11  | test11        |       | ON     |          | 10       |            |
| test2   | test2         |       | ON     |          | 10       |            |
| test3   | test3         |       | ON     |          | 10       |            |
| test4   | test4         |       | ON     |          | 10       |            |
| test5   | test5         |       | ON     |          | 10       |            |
| test6   | test6         |       | ON     |          | 0        | +          |
| test7   | test7         |       | ON     |          | 0        | +          |
| test8   | test8         |       | ON     |          | 0        | +          |
|         |               |       | ALL:14 | FULL:7   | LIMIT:3  | TOTAL:10   |
+---------+---------------+-------+--------+----------+----------+------------+
```

**Reply by ttm-alexeypolunin** [reply]
Ошибка `License format error ... No host in license to check` — это нарушение формата файла лицензии.

Некоторые хостеры предоставляют софт для редактирования файлов через интерфейс и этот софт иногда может добавлять нечитаемые символы в файл, которые нарушают его целостность.

Более того, дальнейшее редактирование файла через консоль эти баги не чинит, тк консоль тоже не видит некоторые спецсимволы. Только полное удаление файла и создание его заново.

**Записывайте файлы лицензии только непосредственно из консоли!**

**Comment by dmbrtlt** [comment]
Спасибо, дело было в том что не из консоли записывал. Проверил тест. Оплатил лицензию, как скоро проводится оплата и лицензия начинает действовать?

**Reply by ttm-alexeypolunin** [reply]
Уже, закрывающие отправляются на email. также можно отправить нам запрос в Диадок и мы будем дублировать туда.

Срок лицензии начнет отсчитываться в момент когда вы нажмете Активировать на сервере лицензий.
